src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -326,6 +329,7 @@ public void Reset()
         {
             _onStarting?.Clear();
             _onCompleted?.Clear();
+            _routeValues?.Clear();
```

Comment:
@rynowak and @JamesNK this opens the door to re-using RouteValues but we'd need to make changes to routing to avoid reassiging this field and instead just adding values. It doesn't need to happen in this PR though but give it some thought.

Reply:
Reusing route values will involves some surgery to replace/reuse the underlying array. The 'slots' optimization that's use to build the route values was a huge perf win, and that likely is more impactful is saving an alloc of the wrapper object. 

Put another way, building the dictionary using `Add` and other methods that set values in the dictionary is slow. Precomputing which indexes the values belong in is fast. So if we want to look at reusing RVD, we also have to look at manipulating its guts directly.

Reply:
It also means that CandidateState needs to expose the slots rather than the RVD https://github.com/aspnet/AspNetCore/blob/2b2aeee9f869ec43f65c188a66d9285858823628/src/Http/Routing/src/Matching/CandidateState.cs#L53. That complicates the API a bit.

Reply:
RVD has a constructor to wrap the underlying array. RVD could be extended to provide a way to get the underlying array instance.

Reply:
It would only need to allocate if there were complex segments or constraints:
 https://github.com/aspnet/AspNetCore/blob/2b2aeee9f869ec43f65c188a66d9285858823628/src/Http/Routing/src/Matching/DfaMatcher.cs#L163

https://github.com/aspnet/AspNetCore/blob/2b2aeee9f869ec43f65c188a66d9285858823628/src/Http/Routing/src/Matching/DfaMatcher.cs#L153

It seems like we need to pass RouteValues to legacy APIs that have them in the signature. Complex segments could be fixed it we cared but it's likely they are uncommon enough to not worry about (this is an internal API).

`IRouteConstraint` takes `RouteValues` as well (and has an `IRouter` in the signature....) so we need to allocate there.

In the other cases, we can make a new API like `FromArray(slots)` but an instance version (it would need to deal with edge cases though). Anyhow, thats future work.